### DISCUSSION TITLE
Re-using field from other object instance where the name of the fields match

### DISCUSSION THREAD

**Discussion by nandorholozsnyak** [instancio/instancio#890] [Q&A]
Hello there,

On Twitter a few days ago I asked a question, and I was routed to here, now I had the time to create an example for this question or use case. I went through the documentation and the different selectors, unfortunately I did not find anything related to this topic.

The use case is the following. I have two different types where one types lists a few attributes, but this type can not/is just no used in the other type as an embedded one, but its fields is replicated in the other type.

For example:

First type:
```java
record Address(String country, String postCode, String county, String city, String addressLine1, String addressLine2) {
}
```

Second type: 
```java
record OrderEntity(String orderId, BigDecimal price, String customerId, String country, String postCode, String county,
                   String city, String addressLine1, String addressLine2, LocalDateTime creationDate) {
}
```

If I create an instance of the Address type with Instancio just with a one liner, or with two lines in case of a shiny custom generator, can I somehow create the OrderEntity with Instancio as well where I do not have to use the `.set(field(` construction to set all fields to the Address attributes?

```java
    public static void main(final String[] args) {
        final var address = Instancio.create(Address.class);
        //Avoid doing this
        final var orderEntity = Instancio.of(OrderEntity.class)
            .set(field(OrderEntity::country), address.country())
            .set(field(OrderEntity::postCode), address.postCode())
            .set(field(OrderEntity::county), address.county())
            .set(field(OrderEntity::city), address.city())
            .set(field(OrderEntity::addressLine1), address.addressLine1())
            .set(field(OrderEntity::addressLine2), address.addressLine2())
            .create();
        // Do something like this to spare the lot of field selectors.
        // But that does not work of course
//        final var anotherOrderEntity = Instancio.of(OrderEntity.class)
//            .set(?whatToUseHere?(OrderEntity.class), address)
//            .create();
        System.out.println(address);
        System.out.println(orderEntity);
    }
```

Does Instancio support this use case, if yes how? 

Thanks in advance.

**Comment by armandino** [comment]
Hey @nandorholozsnyak! Unfortunately, nothing out of the box allows automatically copying fields from one class to another based on field name. This would normally require some sort of manual mapping, as in your example.

Having said that, it's possible to do this using a little helper method. The idea is to find matching fields that exist in both classes, get the field value via reflection, and `set` the value using a selector created from the field:

```java
@Test
void example() {
    final Address address = Instancio.create(Address.class);
    final InstancioApi<OrderEntity> api = Instancio.of(OrderEntity.class);
    
    copyFields(api, address);

    final OrderEntity orderEntity = api.create();

    assertThat(orderEntity.addressLine1()).isEqualTo(address.addressLine1());
    // etc...
}

private static void copyFields(final InstancioApi<OrderEntity> api, final Address address) {
    Stream.of(OrderEntity.class.getDeclaredFields()).forEach(orderField -> {
        try {
            Field addressField = Address.class.getDeclaredField(orderField.getName());
            addressField.setAccessible(true);
            api.set(field(OrderEntity.class, addressField.getName()), addressField.get(address));
        } catch (NoSuchFieldException | IllegalAccessException e) {
            // ignore
        }
    });
}
```

It looks a bit hacky, but If the classes are large and involve many fields, it might be worth it. With a bit more work, it's also possible to make the `copyFields()` method generic, so it can work with any classes.

There might be a better approach, but I can't think of one right now. I'll follow up if I come up with anything.

**Reply by foaw** [reply]
You may also want to look into [MapStruct](https://github.com/mapstruct/mapstruct). It has its own peculiarities, but for simple cases it is generally a working tool.

**Reply by nandorholozsnyak** [reply]
Thank you so much @armandino, I guess a selector could be built for this purpose, I do not say I would you it on a daily basis but on an older legacy project we have so much of these normalized and non normalized class setups.

I was thinking on a contribution just do not know if this would make sense on the long run for Instancio. I guess I can implement it on my own and if it works contribute it to the project.

---
@foaw yeah I do use it a lot, here maybe a target mapping would be appropriate, but I do not make these mappers as test utilities, might make sense tough.

**Reply by armandino** [reply]
Thanks @nandorholozsnyak! Appreciate your offer to submit a PR. I don't know how common this use case and if it's something the core library should provide. If you do implement it, please let me know. I'd be very curious to check it out, especially what the API for this might look like.
